File: _skels/next-js-ddd-skel/src/catalog/repository.js
/**
 * CatalogRepository contract.
 *
 *   list()                          -> CatalogItem[]
 *   get(id)                         -> CatalogItem | null
 *   create({ name, description, price, category, available }) -> CatalogItem
 *
 * CatalogItem rows match the Drizzle schema:
 *   { id, name, description, price, category, available }
 */

const { assertHasMethods } = require('../shared/repository');

const REQUIRED_METHODS = ['list', 'get', 'create'];

function assertCatalogRepository(repo) {
  assertHasMethods(repo, REQUIRED_METHODS, 'CatalogRepository');
}

module.exports = { assertCatalogRepository, REQUIRED_METHODS };

File: _skels/rust-actix-ddd-skel/src/catalog/repository.rs
//! `CatalogRepository` — storage abstraction for catalog items.

use async_trait::async_trait;

use crate::shared::DomainError;

pub use crate::entities::catalog_item::Model as CatalogItem;

#[derive(Debug, Clone)]
pub struct NewCatalogItem {
    pub name: String,
    pub description: String,
    pub price: f64,
    pub category: String,
    pub available: bool,
}

#[async_trait]
pub trait CatalogRepository: Send + Sync {
    async fn list(&self) -> Result<Vec<CatalogItem>, DomainError>;
    async fn get(&self, id: i32) -> Result<CatalogItem, DomainError>;
    async fn create(&self, new: NewCatalogItem) -> Result<CatalogItem, DomainError>;
}
